src/map/skill.cpp

```diff
@@ -16018,22 +16022,20 @@ std::shared_ptr<s_skill_unit_group> skill_unitsetting(struct block_list *src, ui
 	case SA_VIOLENTGALE:
 	case SC_CHAOSPANIC:
 	case SOA_TOTEM_OF_TUTELARY:
-	{
-		std::shared_ptr<s_skill_unit_group> old_sg = skill_locate_element_field(src);
-
-		if (old_sg != nullptr)
-		{	//HelloKitty confirmed that these are interchangeable,
-			//so you can change element and not consume gemstones.
-			if ((
-				old_sg->skill_id == SA_VOLCANO ||
-				old_sg->skill_id == SA_DELUGE ||
-				old_sg->skill_id == SA_VIOLENTGALE
-			) && old_sg->limit > 0)
-			{	//Use the previous limit (minus the elapsed time) [Skotlex]
-				limit = old_sg->limit - DIFF_TICK(gettick(), old_sg->tick);
-				if (limit < 0)	//This can happen...
-					limit = skill_get_time(skill_id,skill_lv);
+		if (std::shared_ptr<s_skill_unit_group> old_sg = skill_locate_element_field(src); old_sg != nullptr) {
+			// HelloKitty confirmed that these are interchangeable, so you can change element and not consume gemstones.
+			if (old_sg->limit > 0) {
+				switch( old_sg->skill_id ) {
+					case SA_VOLCANO:
+					case SA_DELUGE:
+					case SA_VIOLENTGALE:
+						// Use the previous limit (minus the elapsed time) [Skotlex]
+						limit = old_sg->limit - DIFF_TICK(gettick(), old_sg->tick);
+						if (limit < 0)	// This can happen...
```

Comment:
This can really happen on AEGIS?
And you get free casting (without gemstones) in such a case too?

Reply:
Yeah, this is wrong - on Aegis it counts the intact units and both the consumption and the starting of the status change only happen if the count is 0.

It can technically happen because units are removed every 100ms on rAthena, so the sc timer could trigger before the units are removed. Not sure if it can happen on Aegis since the interval is 20ms for everything there.

I would see this code more of a performance optimization as there's not much point to go through all the unit placement code when the unit just disappear within 100ms again.